### PR TITLE
Add case stages for use in hardship claims

### PR DESCRIPTION
**Pull request by jsugarman**
#### What
wire up a new case_stages model for use in hardship claims

#### Ticket

[CBO-1187](https://dsdmoj.atlassian.net/browse/CBO-1187)

#### Why
Design and business requirement to present
users with question about the stage which the case
has reached in hardship claims.

The Business does not believe users will know what
case type/scenario to use and could then submit
higher fees. The case workers drafted in to handle
these claims will not be knowledgable enough to discern
if claim is too high and pay it anyway.

#### How
case stages acts as an interface to a case type. case stages
delegate missing methods to case type so can be used in similar
manner to them. case type will be validated to be nil and
any reference to a case type will, transparently, go via the claims
case_stage.

--------

#### TODO (wip)

 - [x] separate migration and release
 - [x] amend unit tests
 - [x] amend feature test

### CODE REVIEW

File: app/validators/claim/base_claim_validator.rb [hunk 1]
```diff
@@ -64,6 +64,11 @@ def validate_case_type
     validate_presence(:case_type, 'blank')
   end
 
+  # must be absent
+  def validate_stage_type
```
Comment (jsugarman) [inline app/validators/claim/base_claim_validator.rb]: amend to `validate_case_stage` and spec

Comment (jsugarman) [reply app/validators/claim/base_claim_validator.rb]: removed as not needed, really?!

File: features/step_definitions/advocate_hardship_claim_steps.rb [hunk 2]
```diff
@@ -12,24 +12,41 @@
...
+
+Then("I should see the offence details {string}") do |details|
+    expect(page).to have_content('Offence')
+
+    within '.basic-fee' do
+      details.split(',').map(&:strip).each do |detail|
+        expect(page).to have_content(detail)
+      end
+    end
+end
```
Comment (jsugarman) [inline features/step_definitions/advocate_hardship_claim_steps.rb]: line break

File: spec/models/case_stage_spec.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
+RSpec.describe CaseStage, type: :model do
+  subject(:case_stage) { create(:case_stage) }
+
+  it { is_expected.to belong_to(:case_type) }
+
+  # should delegate stuff?! to case type
+  # TODO: write custom matcher for delegate_missing_to rails helper or spec individually
```
Comment (jsugarman) [inline spec/models/case_stage_spec.rb]: TODO

File: app/views/external_users/claims/case_details/_case_type_fields.html.haml [hunk 4]
```diff
@@ -6,8 +6,7 @@
         = t('.stage_type_hint')
       = validation_error_message(@error_presenter, :case_type)
 
-    = f.select :case_type_id, @case_type_stages.map{ |ct| [ct.name, ct.id, { data: { 'is-fixed-fee': ct.is_fixed_fee?, 'requires-cracked-dates': ct&.requires_cracked_dates?, 'requires-retrial-dates': ct&.requires_retrial_dates?, 'requires-trial-dates': ct&.requires_trial_dates? } }] }, { include_blank: ''.html_safe }, { class: 'form-control fx-autocomplete', id: 'case_type', 'aria-label': t('.stage_type') }
-
+    = f.select :case_stage_id, @case_stages.map{ |cs| [cs.description, cs.id, { data: { 'is-fixed-fee': cs.is_fixed_fee?, 'requires-cracked-dates': cs&.requires_cracked_dates?, 'requires-retrial-dates': cs&.requires_retrial_dates?, 'requires-trial-dates': cs&.requires_trial_dates? } }] }, { include_blank: ''.html_safe }, { class: 'form-control fx-autocomplete', id: 'case_type', 'aria-label': t('.stage_type') }
```
Comment (colinbruce) [inline app/views/external_users/claims/case_details/_case_type_fields.html.haml:9]: This seems like a good candidate to pull out to a presenter, but MVP! :)

File: spec/support/validation_helpers.rb [hunk 5]
```diff
@@ -16,12 +16,12 @@ def create_and_submit_claim(claim)
 
   def should_error_with(record, field, message)
     expect(record).not_to be_valid
-    expect(record.errors[field]).to include(message)
+    expect(record.errors[field]).to include(message), "expected #{field} to have the error #{message}, but had #{record.errors[field] || 'none'}"
```
Comment (colinbruce) [inline spec/support/validation_helpers.rb:19]: Nice 👍

**Review by colinbruce — APPROVED** [review]
LGTM :+1:

File: app/views/external_users/claims/case_details/_case_type_fields.html.haml (hunk 4 above)
Comment (jsugarman) [reply app/views/external_users/claims/case_details/_case_type_fields.html.haml:9]: yes, indeed, peter already highlighted the length of the input text

**Pull request merged**